### PR TITLE
ExtendedBuffer Unit Tests & Source code bug fixes/adding of comments

### PR DESCRIPTION
**Pull request by serpens24**
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [X] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

Unit testing of ExtendedBuffer.java
Added comments to ExtendedBuffer.java to document how the clas works.

### Alternate Designs

nil

### Why Should This Be In Core?

unit testing improves code quality, as does good documentation of code

### Benefits

unit testing improves code quality, as does good documentation of code

### Possible Drawbacks

nil
### Verification Process

Tests all pass

### CODE REVIEW

**Comment by serpens24** [comment]
Apologies for the multiple commits ....  Despite tests passing in dev, committing to Github was resulting in errors in several tests. In the end needed to change how exceptions were being tested in trange checking.

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+     */
+    @Test
+    public void testDefaultConstructor() throws Exception {
+        System.out.println("ExtendedBufferNGTest.testDefaultConstructor");
+
+        final ExtendedBuffer localBuffer = new ExtendedBuffer(size);
+        final Field bufferSize = ExtendedBuffer.class.getDeclaredField("bufferSize");
+        bufferSize.setAccessible(true);
+        assertEquals(size, (int)bufferSize.get(localBuffer));
+        assertEquals(localBuffer.getAvailableSize(), 0);
```
Comment (arcturus2) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java:58]: I'm ok with the use of reflection here but the other option would be to add a getter for the bufferSize.

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBuffer.java [hunk 2]
```diff
@@ -165,29 +236,73 @@ public int read(byte[] b, int off, int len) throws IOException {
...
         @Override
         public void write(byte[] b, int off, int len) throws IOException {
 
+            // Offset into source array must reside within array
+            if (off >= b.length) {
+                throw new IOException("Source offset outside of range");
+            }
+            
+            // Check if off + len would spill over the back of the source array and adjust len accordingly
+            len = Math.min(len, b.length - off);
```
Comment (arcturus2) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBuffer.java:292]: 👍

**Review by arcturus2 — APPROVED** [review]

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+/**
+ *
+ * @author serpens24
+ */
+public class ExtendedBufferNGTest {
+    
+    final private int size = 10;
```
Comment (antares1470) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java:25]: minor but can you swap `private` and `final` around

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java (hunk 1 above)
Comment (antares1470) [reply CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java:58]: yeah I think a protected getter might be a better option in this instance

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        final ExtendedBuffer buffer = new ExtendedBuffer(size);
+        final OutputStream outputStream = buffer.getOutputStream();
+        final byte[] bytes = "ABCDE".getBytes();
+        
+        try {
+            // Check write doesnt permit offset to be outside of source
+            // array size
+            try {
+            outputStream.write(bytes, 6, 1);
+            fail("Exception not thrown");
```
Comment (antares1470) [inline CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java:115]: are you able to fix the formatting here?

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java (hunk 1 above)
Comment (serpens24) [reply CoreUtilities/test/unit/src/au/gov/asd/tac/constellation/utilities/stream/ExtendedBufferNGTest.java:58]: Yep thats an easy add.

**Comment by serpens24** [comment]
Delete an antares comment as it wa a duplicate of arcturus original "reflection" comment, the tool wouldnt let me acknowledge it.

**Comment by serpens24** [comment]
Need to work out why a test is now failing and if thats in extended buffer

**Review by antares1470 — APPROVED** [review]

**Pull request merged**